fuel-vm/src/tests/upgrade.rs

```diff
@@ -0,0 +1,418 @@
+
+        // Then
+        assert!(client
+            .as_mut()
+            .state_transition_bytecodes_versions_mut()
+            .contains_key(&expected_version));
+    }
+
+    #[test]
+    fn transact_with_zero_gas_price_doesnt_affect_change_output() {
+        let state_transition_hash = [1; 32].into();
+        let mut client = Interpreter::<_, Upgrade>::with_storage(
+            valid_storage(state_transition_hash, vec![]),
+            InterpreterParams::default(),
+        );
+
+        // Given
+        let gas_price = 0;
+        client.set_gas_price(gas_price);
+        let tx = valid_transaction(state_transition_hash)
+            .into_ready(gas_price, &GasCosts::default(), &Default::default())
+            .expect("failed to generate ready tx");
+
+        // When
+        let state = client.transact(tx).expect("failed to transact");
+
+        // Then
+        let Output::Change {
+            amount, asset_id, ..
+        } = state.tx().outputs()[0]
+        else {
+            panic!("expected change output");
+        };
+        assert_eq!(amount, AMOUNT);
+        assert_eq!(asset_id, AssetId::BASE);
+    }
+
+    #[test]
+    fn transact_with_non_zero_gas_price_affects_change_output() {
+        let state_transition_hash = [1; 32].into();
+        let mut client = Interpreter::<_, Upgrade>::with_storage(
+            valid_storage(state_transition_hash, vec![]),
+            InterpreterParams::default(),
+        );
+
+        // Given
+        let gas_price = 1;
+        client.set_gas_price(gas_price);
+        let tx = valid_transaction(state_transition_hash)
+            .into_ready(gas_price, &GasCosts::default(), &Default::default())
+            .expect("failed to generate ready tx");
+
+        // When
+        let state = client.transact(tx).expect("failed to transact");
+
+        // Then
+        let Output::Change {
+            amount, asset_id, ..
+        } = state.tx().outputs()[0]
+        else {
+            panic!("expected change output");
+        };
+        assert_eq!(amount, AMOUNT - 1);
+        assert_eq!(asset_id, AssetId::BASE);
+    }
+
+    #[test]
+    fn transact_fails_for_unknown_hash() {
+        let known_state_transition_hash = [1; 32].into();
+        let mut client = Interpreter::<_, Upgrade>::with_storage(
+            valid_storage(known_state_transition_hash, vec![]),
+            InterpreterParams::default(),
+        );
+
+        // Given
+        let unknown_state_transition_hash = [2; 32].into();
+        let tx = valid_transaction(unknown_state_transition_hash).test_into_ready();
+
+        // When
+        let result = client.transact(tx).map(|_| ());
+
+        // Then
+        assert_eq!(
+            Err(InterpreterError::Panic(
+                PanicReason::UnknownStateTransactionBytecodeHash
+            )),
+            result
+        );
+    }
+
+    #[test]
+    fn transact_fails_when_try_to_override_state_bytecode() {
```

Comment:
This case is confusing me. Isn't the point of the upgrade to override some previous state? 

Other than the first time it's run, under what circumstances will this not fail?

Reply:
Same question for the consensus params.

Reply:
Each version contains only one associated consensus parameter or state transition function. If it was set before, it is not allowed to be overridden for the same version. You can only increase the version and set it for a new one.